src/persist/src/indexed/unsealed.rs

```diff
@@ -314,20 +315,77 @@ pub struct UnsealedSnapshot {
 }
 
 impl Snapshot<Vec<u8>, Vec<u8>> for UnsealedSnapshot {
-    fn read<E: Extend<((Vec<u8>, Vec<u8>), u64, isize)>>(
-        &mut self,
-        buf: &mut E,
-    ) -> Result<bool, Error> {
-        if let Some(batch) = self.batches.pop() {
-            let batch = batch.recv()?;
-            let updates = batch
-                .updates
-                .iter()
-                .filter(|(_, ts, _)| self.ts_lower.less_equal(ts) && !self.ts_upper.less_equal(ts))
-                .map(|((key, val), ts, diff)| ((key.clone(), val.clone()), *ts, *diff));
-            buf.extend(updates);
+    type Iter = UnsealedSnapshotIter;
+
+    fn into_iters(self, num_iters: usize) -> Vec<Self::Iter> {
+        let mut iters = Vec::with_capacity(num_iters);
+        iters.resize_with(num_iters, || UnsealedSnapshotIter {
+            ts_lower: self.ts_lower.clone(),
+            ts_upper: self.ts_upper.clone(),
+            current_offset: 0,
+            current_batch: None,
+            batches: VecDeque::new(),
+        });
+        // TODO: This should probably distribute batches based on size, but for
+        // now it's simpler to round-robin them.
+        for (i, batch) in self.batches.into_iter().enumerate() {
+            let iter_idx = i % num_iters;
+            iters[iter_idx].batches.push_back(batch);
+        }
+        iters
+    }
+}
+
+/// An [Iterator] representing one part of the data in a [UnsealedSnapshot].
+//
+// This intentionally stores the batches as a VecDeque so we can return the data
+// in roughly increasing timestamp order, but it's unclear if this is in any way
+// important.
+#[derive(Debug)]
+pub struct UnsealedSnapshotIter {
+    /// A closed lower bound on the times of contained updates.
+    ts_lower: Antichain<u64>,
+    /// An open upper bound on the times of the contained updates.
+    ts_upper: Antichain<u64>,
+
+    current_offset: usize,
+    current_batch: Option<Result<Arc<BlobUnsealedBatch>, Error>>,
+    batches: VecDeque<Future<Arc<BlobUnsealedBatch>>>,
+}
+
+impl Iterator for UnsealedSnapshotIter {
+    type Item = Result<((Vec<u8>, Vec<u8>), u64, isize), Error>;
+
+    fn next(&mut self) -> Option<Self::Item> {
```

Comment:
Ooof this whole impl (and the corresponding one for trace) is rough. Taking suggestions for how this could be better written if you have them

Reply:
You could turn `current_offset`/`current_batch` into just `current_batch` which is itself an `Iterator`. That way, you would factor the last part of the method out into something else. I mean this part:
```rust
if self.current_offset >= current_batch.updates.len() {
    self.current_batch = None;
    continue;
}
let (kv, ts, diff) = &current_batch.updates[self.current_offset];
self.current_offset += 1;
if !(self.ts_lower.less_equal(ts) && !self.ts_upper.less_equal(ts)) {
    continue;
}
return Some(Ok((kv.clone(), *ts, *diff)));
```

The "main" code here would become a thing that pops of iterators until they're all depleted. Not sure if it helps much, though.

Reply:
That sounds promising but since this is an Arc<BlobUnsealedBatch> we get an iterator with a lifetime and I'm not quite sure what to do with that?

Reply:
Neither do I, but I have a feeling that GATs would solve it? :thinking:

Reply:
Realized this morning that I could clone the updates vec (or collect it into a VecDeque) and store that as current_batch. It's not as "lazy" as what is here but I'd guess the important users of snapshots will always read the entirety of the iterators.

Reply:
Pushed this approach, seems much cleaner. It also is less lazy in the sense of cloning data that is rejected by the filter, but that filter is expected to be pretty permissive in practice, so I'm not terribly worried about the either. It really seems like the logic of these two impls should be extractable into a single place, but the couple of things I tried made it less clear

Reply:
this is fairly minor but you could push the filter into the cloning step I think with something like:

```
diff --git a/src/persist/src/indexed/unsealed.rs b/src/persist/src/indexed/unsealed.rs
index 0e472f431..9a275d632 100644
--- a/src/persist/src/indexed/unsealed.rs
+++ b/src/persist/src/indexed/unsealed.rs
@@ -358,10 +358,7 @@ impl Iterator for UnsealedSnapshotIter {
         loop {
             if !self.current_batch.is_empty() {
                 let (kv, ts, diff) = self.current_batch.pop().unwrap();
-                if self.ts_lower.less_equal(&ts) && !self.ts_upper.less_equal(&ts) {
-                    return Some(Ok((kv, ts, diff)));
-                }
-                continue;
+                return Some(Ok((kv, ts, diff)));
             } else {
                 // current_batch is empty, find a new one.
                 let b = match self.batches.pop_front() {
@@ -370,9 +367,19 @@ impl Iterator for UnsealedSnapshotIter {
                 };
                 match b.recv() {
                     Ok(b) => {
+                        let ts_lower = self.ts_lower.clone();
+                        let ts_upper = self.ts_upper.clone();
                         // Reverse the updates so we can pop them off the back
                         // in roughly increasing time order.
-                        self.current_batch.extend(b.updates.iter().rev().cloned());
+                        self.current_batch.extend(
+                            b.updates
+                                .iter()
+                                .rev()
+                                .filter(|(_, ts, _)| {
+                                    ts_lower.less_equal(ts) && !ts_upper.less_equal(ts)
+                                })
+                                .cloned(),
+                        );
                         continue;
                     }
                     Err(err) => return Some(Err(err)),
```

its not clear the additional upper and lower clones are worth it for pushing the filter here though

Reply:
Lovely! We don't even need to clone the antichains if we borrow them into antichainrefs